src/terms/xc.jl

```diff
@@ -51,117 +51,41 @@ end
     density = LibxcDensity(basis, max_ρ_derivs, ρ, ρspin)
 
     potential = [zeros(T, basis.fft_size) for _ in 1:n_spin]  # TODO CPU arrays
-    zk = zeros(T, basis.fft_size)  # Energy per unit particle
-    E = zero(T)
-    for xc in term.functionals
-        # Evaluate the functional and its first derivative (potential)
-        terms = evaluate(xc, density; zk=zk)
-
-        # Add energy contribution
-        dVol = basis.model.unit_cell_volume / prod(basis.fft_size)
-        E += sum(terms.zk .* ρ.real) * dVol
-
-        # Add potential contributions Vρ -2 ∇⋅(Vσ ∇ρ)
-        for σ in 1:n_spin
-            potential[σ] .+= @view terms.vrho[σ, :, :, :]
-        end
+    terms = evaluate(term.functionals, density)
+
+    # Energy contribution (zk == energy per unit particle)
+    dVol = basis.model.unit_cell_volume / prod(basis.fft_size)
+    E = sum(terms.zk .* ρ.real) * dVol
+
+    # Map from the cartesian index into the two spin indices of the contracted
+    # density gradient σ_{αβ} to the convention used in libxc.
+    tσ = libxc_spinindex_σ
+
+    # Potential contributions Vρ -2 ∇⋅(Vσ ∇ρ)
+    @views for s in 1:n_spin
+        potential[s] .+= terms.vrho[s, :, :, :]
 
         if haskey(terms, :vsigma)  # Need gradient correction
-            # TODO Potential to save some FFTs for spin-polarised calculations:
-            #      For n_spin == 2 this calls divergence_real 4 times, where only 2 are
-            #      needed (one for potential[1] and one for potential[2])
-            @views for σ in 1:n_spin, τ in 1:n_spin
-                στ = libxc_symmetric_index(σ, τ)
-                # Extra factor (1/2) for σ != τ is needed because libxc only keeps σ_{αβ}
+            potential[s] .+= -2divergence_real(density.basis) do α
+                # Extra factor (1/2) for s != t is needed because libxc only keeps σ_{αβ}
                 # in the energy expression. See comment block below on spin-polarised XC.
-                spinfac = (σ == τ ? 2 : 1)
-                potential[σ] .+=
-                     -spinfac .* divergence_real(
-                        α -> terms.vsigma[στ, :, :, :] .* density.∇ρ_real[τ, :, :, :, α],
-                        density.basis
-                    )
+                sum((s == t ? one(T) : one(T)/2)
+                    .* terms.vsigma[tσ(s, t), :, :, :] .* density.∇ρ_real[t, :, :, :, α]
```

Comment:
Well that's sort of the point of these `tσ` etc functions. They hide the weird index mapping you need to do. I agree it would be nice to hide that better, but at that level (i.e. where you contract and build the potential) I don't think there is much you can do easily without copying the arrays around.